internal/pkg/template/diff/diff.go

```diff
@@ -0,0 +1,28 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package diff provides functionalities to compare two YAML documents.
+package diff
+
+import (
+	"gopkg.in/yaml.v3"
+)
+
+// Node represents a segment on a difference between two YAML documents.
+type Node struct {
+	key      string
+	children map[string]*Node // A list of non-empty pointers to the children nodes.
+
+	oldValue *yaml.Node // Only populated for a leaf node (i.e. that has no child node).
```

Comment:
Dumb question: Why do we only check values for child nodes?

Reply:
It is not necessary to store the yaml content for a branch node. The YAML content under the branch node can be reconstructed using its `key` and all of its `children`.

This probably will be clearer if you see the implementation code.